site/js/modules/webui/render.js

```diff
@@ -1,35 +1,104 @@
+import * as Autograder from '../autograder/base.js';
+
 import * as Assignment from './assignment.js';
 import * as Routing from './routing.js';
 import * as Util from './util.js';
 
-function makeCardObject(type = 'unknown', text = '', link = '#') {
-    return {
-        type: type,
-        text: text,
-        link: link,
-    };
-}
+class Card {
+    constructor(
+        type = 'unknown', text = '', link = '#',
+        {
+            minServerRole = Autograder.Users.SERVER_ROLE_UNKNOWN,
+            minCourseRole = Autograder.Users.COURSE_ROLE_UNKNOWN,
+            courseId = undefined
+        } = {}) {
+        // An optional card type that is added to the HTML class list.
+        this.type = type;
 
-function card(card = {type: 'unknown', text: '', link: '#'}) {
-    return `
-        <div class='card card-${card.type} secondary-color drop-shadow'>
-            <a href='${card.link}' alt='${card.text}'>
-                <span>${card.text}</span>
-            </a>
-        </div>
-    `;
+        // The display test of the card.
+        this.text = text;
+
+        // Routes to this link when the card is clicked.
+        // All routes must start with a '#'.
+        this.link = link;
+
+        // The minimum server role a user needs to have to view this card.
+        this.minServerRole = minServerRole;
+
+        // The minimum course role a user needs to have to view this card.
+        this.minCourseRole = minCourseRole;
+
+        // The course that is used for the course role check.
+        this.courseId = courseId;
+
+        this.validate();
+    }
+
+    validate() {
+        if (!this.link.startsWith('#')) {
+            console.error(`A card link must start with a '#': '${this.link}'.`);
+        }
+
+        if (!Number.isInteger(this.minServerRole)) {
+            console.error('A card must have an integer value for the min server role.');
+        }
+
+        if (!Number.isInteger(this.minCourseRole)) {
+            console.error('A card must have an integer value for the min course role.');
+        }
+    }
+
+    toHTML() {
+        return `
+            <div class='card card-${this.type} secondary-color drop-shadow'>
+                <a href='${this.link}' alt='${this.text}'>
+                    <span>${this.text}</span>
+                </a>
+            </div>
+        `;
+    }
+
+    hide(context) {
```

Comment:
The name makes these seem like an action
(like call this method to hide a card).
Maybe instead have something like "isHidden()"?